src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+                        }
+                        catch (RequestFailedException ex) when (ex.ErrorCode == BlobErrorCode.BlobNotFound)
+                        {
+                            _logger.LogError(new UnclaimedPartitionException($"Partition {partition} has never been claimed"));
+                            _logger.LogMetric(EventMetrics.PartitionLastClaimedDelay(partition), DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
+                        }
+                    }
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private Task RenewOwnership(CancellationToken ct)
+        {
+            Task.Run(async () =>
+            {
+                while (!ct.IsCancellationRequested)
+                {
+                    foreach (var partition in _ownedPartitions)
+                    {
+                        _logger.LogTrace($"Renewing ownership for partition {partition}");
+                        BlobClient ownershipClient = _containerClient.GetBlobClient($"ownedpartitions/{partition}");
+                        var leaseClient = ownershipClient.GetBlobLeaseClient(_processorId);
+                        await leaseClient.RenewAsync(default, ct);
+
+                        // upload to update last modified timestamp
+                        BlobUploadOptions blobUploadOptions = new BlobUploadOptions()
+                        {
+                            Conditions = new BlobRequestConditions()
+                            {
+                                LeaseId = _processorId,
+                            },
+                        };
+                        await ownershipClient.UploadAsync(BinaryData.FromString(string.Empty), blobUploadOptions);
+                    }
+
+                    await Task.Delay(20000, ct);
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private async Task StartAssignedPartitionProcessor(CancellationToken ct, List<int> ownedPartitions)
+        {
+            var maximumBatchSize = 100;
+
+            string[] partitions = new string[ownedPartitions.Count];
+            for (var i = 0; i < ownedPartitions.Count; i++)
+            {
+                partitions[i] = ownedPartitions.ElementAt(i).ToString();
+            }
+
+            var eventHubName = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubName, nameof(_eventHubClientOptions.EventHubName));
+            var eventHubNamespaceFQDN = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubNamespaceFQDN, nameof(_eventHubClientOptions.EventHubNamespaceFQDN));
+            var eventHubConsumerGroup = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubConsumerGroup, nameof(_eventHubClientOptions.EventHubConsumerGroup));
+            var tokenCrendential = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubTokenCredential, nameof(_eventHubClientOptions.EventHubTokenCredential));
+
+            var processor = new AssignedPartitionProcessor(
```

Comment:
From what I can see in the code, an instance of `AssignedPartitionProcessor` is created every 5 seconds - Is this correct?

Line 332 is in a loop that executes every 5 seconds and calls `RunAndRenew()`
Line 309 in `RunAndRenew()` calls `StartAssignedPartitionProcessor()` which creates a new `AssignedPartitionProcessor`.

Reply:
Nah the AssignedPartitionProcessor is only created once. 

We essentially go to RunProcessorAndRenewPartitionOwnership which creates StartAssignedPartitionProcessor and AssignedPartitionProcessor but we basically just await in there until we receive a cancellation token.

```
                await assignedPartitionProcessor.StartProcessingAsync(ct);
                await Task.Delay(Timeout.Infinite, ct);
```